### PR TITLE
[Dy2stat]Support buffers and register_buffer in Layer

### PR DESCRIPTION
**Pull request by Aurelius84**
### PR types

New features

### PR changes

APIs

#### For imperative
image

#### For @declaratvie
image

**Case 样例：**
```python
class Net(Layer):
    def __init__(self, batch_size, hidden_size, seq_len_mask):
        self.gru_unit = DynamicGRU(...)
        # RNN层的起始step的h_0值
        self.h_0 = to_variable(np.zeros([batch_size, hidden_size]).astype('float32'))
        # mask矩阵
        self.mask = to_variable(seq_len_mask) # [batch_size, max_seq_len]
    
    def forward(self, inputs):
        gru_out = self.gru_unit(inputs, self.h_0)
        mask_out = gru_out * self.mask
        return mask_out
```

+ Add `_buffers` in `dygraph.Layer` to store VarBase  created in `__init__ `
+ Support `register_buffer` to allow user to add persistent var into state_dict
+ Support to save varBase in `self._buffers` into inference model in Dy2static

```python
# non_persistable buffer
with fluid.dygraph.guard():
    net = fluid.Layer()
    var1 = to_variable(np.zeros([1]))

    net.register_buffer("buffer_name", var1, persistable=False)
    self.assertEqual(len(net.buffers()), 1)
    self.assertEqual(len(net.state_dict()), 0)

# switch persistable of buffer
with fluid.dygraph.guard():
    net = fluid.Layer()
    var1 = to_variable(np.zeros([2, 3]))
    var2 = to_variable(np.zeros([3, 2]))
    net.register_buffer("buffer_var1", var1)
    net.register_buffer("buffer_var2", var2, persistable=False)

    self.assertEqual(len(net.state_dict()), 1)
    self.assertEqual([name for name, _ in net.state_dict().items()],
                     ["buffer_var1"])

    # load state_dict
    net_load = fluid.Layer()
    var = to_variable(np.ones([2, 3]))
    net_load.register_buffer("buffer_var1", var)
    net_load.load_dict(net.state_dict())
```

+ Add **LAC model** for dy2stat unittest
+ Polish warning by replace `logger.info` with `warning.warn` to avoid repeatedly printing log.

**Before:**
```
2020-06-08 14:55:06,488-INFO: The decorator 'declarative' doesn't work when setting ProgramTranslator.enable=False. We will just return dygraph output.
[train] step = 0, loss = 111.852859, P: 0.179894, R: 0.194286, F1: 0.186813, elapsed time 0.106487
2020-06-08 14:55:06,595-INFO: The decorator 'declarative' doesn't work when setting ProgramTranslator.enable=False. We will just return dygraph output.
2020-06-08 14:55:06,664-INFO: The decorator 'declarative' doesn't work when setting ProgramTranslator.enable=False. We will just return dygraph output.
2020-06-08 14:55:06,744-INFO: The decorator 'declarative' doesn't work when setting ProgramTranslator.enable=False. We will just return dygraph output.
```
**After:**
```
paddle/python/paddle/fluid/dygraph/jit.py:158: UserWarning: The decorator 'declarative' doesn't work when setting ProgramTranslator.enable=False. We will just return dygraph output.
[train] step = 0, loss = 111.852859, P: 0.179894, R: 0.194286, F1: 0.186813, elapsed time 0.106487
```

### 文档预览
中文文档PR： https://github.com/PaddlePaddle/FluidDoc/pull/2229

image
image
image

### CODE REVIEW

File: python/paddle/fluid/dygraph/base.py [hunk 1]
```diff
@@ -64,17 +64,35 @@ def program_desc_tracing_guard(enable):
...
+                # but VarBase in self._buffers. In dy2static, we convert all the VarBases from
+                # `self._parameters` and `self._buffers` into framework.Parameter.
+                # Because users can create a VarBase in `__init__` equivalent to a ParamBase like a
+                # `mask` Tensor or `hidden_0` in RNN layers, which is necessary for inferring.
+                # So we convert and save them into inference model.
+                if isinstance(var_base, framework.ParamBase):
+                    new_var = var_base._to_static_var(to_parameter=True)
+                else:
+                    # A VarBase in self._buffers may be passed to initialize multiple sub_layers.
+                    is_created = var_base.name in var_base.block.vars
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/base.py]: Seems `is_created` is used only once, you can inline the line 82 into line 83, such as

if var_base.name in var_base.block.vars

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_lac.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+            origin_mode=origin_mode)
+
+        self.size = size
+        self.h_0 = h_0
+        self.is_reverse = is_reverse
+
+    def forward(self, inputs):
+        inputs = to_variable(inputs)
+        hidden = to_variable(self.h_0)
+        hidden.stop_gradient = True
```
Comment (liym27) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_lac.py]: why call `to_variable` here?  In the original model, `to_variable` is not called.
And why set stop_gradient = True?

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_lac.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+                # avoid this bug, and working on fixing it in next PR.
+                j = i + 0
+
+            if i == 0:
+                h_0 = to_variable(self.h_0)
+            else:
+                h_0 = hidden
+
+            input_ = fluid.layers.slice(
+                inputs, axes=[1], starts=[j], ends=[j + 1])
```
Comment (liym27) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_lac.py]: Better add a code comment to explain why not use `inputs[:,i:i+1,:]`

File: python/paddle/fluid/dygraph/base.py [hunk 4]
```diff
@@ -64,17 +64,35 @@ def program_desc_tracing_guard(enable):
...
+                    if is_created:
+                        new_var = var_base.block.vars[var_base.name]
+                        if -1 in new_var.shape:
+                            logging.warn(
+                                "The VarBase [{}] has been converted into Parameter in @declarative mode, "
+                                "but its shape is modified into: {}, we will reset it into {}.".
+                                format(var_base.name, new_var.shape,
```
Comment (liym27) [inline python/paddle/fluid/dygraph/base.py]: Better explain why modify VarBase.shape ~

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py [hunk 5]
```diff
@@ -194,18 +211,20 @@ def from_func_spec(func_spec):
             inputs = func_spec.to_static_inputs(main_program)
 
             # 2. Gets all ParamBases in the function
-            all_parameters = list(func_spec.parameters().values())
+            all_parameters_or_buffers = list(func_spec.parameters().values(
+            )) + list(func_spec.buffers().values())
```
Comment (liym27) [inline python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py]: all_parameters_or_buffers -> all_parameters_and_buffers ?

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py [hunk 6]
```diff
@@ -106,6 +108,21 @@ def parameters(self, include_sublayer=True):
...
+        """
+        buffers = collections.OrderedDict()
+        if self.is_method():
+            if include_sublayer:
+                buffers = self._args[0].buffers()
+                names = [buffer.name for buffer in buffers]
+                buffers = collections.OrderedDict(zip(names, buffers))
+            else:
+                buffers = self._args[0]._buffers
+        return buffers
```
Comment (liym27) [inline python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py:161]: Better more readable code as we discussed offline.

File: python/paddle/fluid/dygraph/base.py (hunk 1 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/base.py]: Thanks，done.

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_lac.py (hunk 3 above)
Comment (Aurelius84) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_lac.py]: Thanks, done.

File: python/paddle/fluid/dygraph/base.py (hunk 4 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/base.py]: Thanks, done.

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py (hunk 5 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py]: Thanks, done.

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py (hunk 6 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py:161]: Thanks, done.

**Comment by Aurelius84** [comment]
> 我们是否需要一个`add_buffered_var`接口，王震说他做量化有个需求，有些变量他有时候想和模型一起存，有时候又不想一起存，在静态图下，这个只要把var设置成persistable，save_inference_model就能存储了，但是动态图下目前不行

已在`layer.Layer`中提供了一个`register_buffer(name, var_base, persistable=True)`的接口，支持动态图下显式将一个varbase添加进buffer是，并同步到state_dict中，支持save/load。

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_lac.py (hunk 2 above)
Comment (Aurelius84) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_lac.py]: 1. why call `to_variable`?
Because in the following code, `hidden` will be assigned a new shape [-1, 128] but it has been modified into `Parameter` in Dy2stat. We don't allow parameter hold shape with -1 which will raise a error if you call `program.clone()`.
Here  I call `to_variable` to make a new var with same data of `self.h_0` to avoid the above problem. Maybe we should give a warning to notice users don't use in the way.

2. why set stop_gradient = True?
Because hidden is modified to Parameter with trainable=True and shall be not updated, so it don't hold gradient data.

**Comment by Aurelius84** [comment]
> LGTM.
> There is a slight difference between Paddle and other framework with regard to this feature,
> `self.tensor = tensor` and `self.register_buffer(tensor)` behaves differently in other framework. The former does not add tensor to `_buffers` while the latter does.

You're right, in this PR, the following interface is same as other framework:
+ `buffers()` : return a list of all buffers from this layer
+ `register_buffer`: add a buffer into the layer.
+ `state_dict`: return all parameters and persistable buffers of this layer

The difference between ours and other framework:
+ PyTorch only support one way to register buffer by calling `register_buffer` explictly.
+ In Paddle, `self.buf_name = var` will add a buffer implictly，because it's necessary for Dy2stat. But because only persistent buffers can be put into `state_dict`, so we make them `persistable=False` created by assigned way to keep consistent with other framework while `save/load`.

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py [hunk 7]
```diff
@@ -96,16 +94,37 @@ def is_method(self):
...
                 names = [p.name for p in params]
                 params = collections.OrderedDict(zip(names, params))
             else:
-                params = self._args[0]._parameters
+                params = layer_instance._parameters
         return params
 
+    def buffers(self, include_sublayer=True):
+        """
+        Returns VarBase buffers of decorated layers. If set `include_sublayer` True,
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py]: 话说文档上我们是不是只向用户暴露了`Variable`和`Parameter`的概念，VarBase和ParamBase是不是没暴露

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py [hunk 8]
```diff
@@ -381,7 +402,7 @@ def func(x):
             dygraph_func
         ), "Input dygraph_func is not a callable in ProgramTranslator.get_output"
         if not self.enable_declarative:
-            logger.info(
+            warnings.warn(
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py:463]: warnings.warn提醒的时候会将代码也贴出来，提示内容会比较多，适用于需要提示用户修改代码的场景，这里其实是告诉用户我们根据当前模式返回了合适的结果，是否还是logging.warning更简洁一些

File: python/paddle/fluid/dygraph/layers.py [hunk 9]
```diff
@@ -414,6 +417,94 @@ def named_sublayers(self,
                         layers_set=layers_set):
                     yield p, l
 
+    def register_buffer(self, name, var_base, persistable=True):
+        """Adds a buffer to the layer.
+
+        This is typically used to register a buffer that should not be
+        considered as a parameter and it will also not be updated by optimizer.
+        The registered buffer is persistable by default, and will be saved into
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/layers.py:426]: `Adds a buffer`? 是否先解释一下这里的buffer是作为和parameter类似的variable概念，是不是容易理解为buffer本来的意思，我读到这里第一反应这里漏掉了一个词，`Adds a buffered variable`，细想才觉得没问题

File: python/paddle/fluid/dygraph/layers.py [hunk 10]
```diff
@@ -414,6 +417,94 @@ def named_sublayers(self,
...
+        state_dict alongside parameters. If set persistable=False, it registers
+        a non-persistable buffer, so that it will not be a part of state_dict.
+
+        Buffers can be accessed as attributes using given names.
+
+        Parameters:
+            name (string): name of the buffer. The buffer can be accessed
+                from this layer using the given name
+            var_base (VarBase): buffer to be registered.
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/layers.py]: 参数名谨慎设计，既然我们不向用户解释什么是`varbase`，那这里是否用`variable`比较合适，类型也应该是Variable

File: python/paddle/fluid/dygraph/layers.py [hunk 11]
```diff
@@ -414,6 +417,94 @@ def named_sublayers(self,
...
+        """
+        Returns an iterator over all buffers in the Layer, yielding tuple of name and varBase.
+
+        Parameters:
+            prefix(str, optional): Prefix to prepend to all buffer names. Default: ''.
+            include_sublayers(bool, optional): Whether include the buffers of sublayers.
+                If True, also include the named buffers from sublayers. Default: True.
+
+        Yields:
+            (string, VarBase): Tuple of name and VarBase
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/layers.py]: 同上，VarBase -> Variable

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py (hunk 7 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py]: Thanks，这里应该写Variable，已修改

File: python/paddle/fluid/dygraph/layers.py (hunk 9 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/layers.py:426]: Thanks，第一句修改为对buffer和parameter之间区别的描述。

File: python/paddle/fluid/dygraph/layers.py (hunk 10 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/layers.py]: Thanks，var_base参数名很不合适，已改为var_base.

File: python/paddle/fluid/dygraph/layers.py (hunk 11 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/layers.py]: Thanks，done.

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py (hunk 8 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py:463]: 这里warning.warn不会输出代码。这里替换logging.warning是因为避免重复输出log，淹没和训练相关的输出。

File: python/paddle/fluid/dygraph/base.py [hunk 12]
```diff
@@ -64,17 +64,26 @@ def program_desc_tracing_guard(enable):
...
-                    var_base.type,
-                    name=var_base.name)
+                # Convert ParamBase into Parameter with same attributes in dy2stat.
+                if isinstance(var_base, framework.ParamBase):
+                    new_var = var_base._to_static_var(to_parameter=True)
+                else:
+                    # Check whether has been created before.
+                    if var_base.name in var_base.block.vars:
+                        new_var = var_base.block.vars[var_base.name]
+                    # Note(Aurelius84): Convert VarBase in self._buffers into Variabe with
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/base.py:79]: `Variabe` -> `Variable`, not important
